File: backend/aisle/ingest/connectors/reddit.py
"""Reddit connector — PRAW, read-only. Pulls both submissions AND comments
(§4 — comments carry the richest reasoning) across the configured
subreddits, searching the configured query terms. Disabled by default in
config/sources.yaml until REDDIT_CLIENT_ID/SECRET are set.
"""
from __future__ import annotations

from datetime import datetime, timezone

from aisle.ingest.connectors._mock_pool import mock_fetch
from aisle.ingest.connectors.base import Connector
from aisle.ingest.schema import RawDoc


class RedditConnector(Connector):
    kind = "reddit"

    def fetch(self, since: datetime | None, limit: int, dry_run: bool = False) -> list[RawDoc]:
        from aisle.settings import get_settings

        settings = get_settings()
        if settings.mock_mode:
            return mock_fetch(self.source_name, since, limit)
        return self._fetch_live(since, limit, dry_run, settings)

    def _fetch_live(self, since, limit, dry_run, settings) -> list[RawDoc]:
        client_id = settings.require("reddit_client_id")
        client_secret = settings.require("reddit_client_secret")

        if dry_run:
            subs = self.config.get("subreddits", [])
            terms = self.config.get("query_terms", [])
            return [
                RawDoc(
                    external_id="dryrun",
                    raw_text=f"[dry-run] would search r/{'+'.join(subs)} for {terms}",
                    author="dry-run",
                    source_name=self.source_name,
                    meta={"dry_run": True},
                )
            ]

        import praw

        reddit = praw.Reddit(
            client_id=client_id,
            client_secret=client_secret,
            user_agent=settings.reddit_user_agent,
        )
        docs: list[RawDoc] = []
        subreddits = self.config.get("subreddits", [])
        terms = self.config.get("query_terms", [])
        include_comments = self.config.get("include_comments", True)

        for sub_name in subreddits:
            subreddit = reddit.subreddit(sub_name)
            for term in terms:
                for submission in subreddit.search(term, sort="new", limit=limit // max(1, len(terms))):
                    posted_at = datetime.fromtimestamp(submission.created_utc, tz=timezone.utc)
                    if since is not None and posted_at <= since:
                        continue
                    docs.append(
                        RawDoc(
                            external_id=submission.id,
                            raw_text=f"{submission.title}\n\n{submission.selftext}",
                            author=str(submission.author) if submission.author else "deleted",
                            source_name=self.source_name,
                            posted_at=posted_at,
                            url=f"https://reddit.com{submission.permalink}",
                            meta={"subreddit": sub_name, "kind": "submission", "score": submission.score},
                        )
                    )
                    if include_comments:
                        submission.comments.replace_more(limit=0)
                        for comment in submission.comments.list():
                            c_posted_at = datetime.fromtimestamp(comment.created_utc, tz=timezone.utc)
                            if since is not None and c_posted_at <= since:
                                continue
                            docs.append(
                                RawDoc(
                                    external_id=comment.id,
                                    raw_text=comment.body,
                                    author=str(comment.author) if comment.author else "deleted",
                                    source_name=self.source_name,
                                    posted_at=c_posted_at,
                                    url=f"https://reddit.com{comment.permalink}",
                                    meta={"subreddit": sub_name, "kind": "comment", "score": comment.score},
                                )
                            )
                    if len(docs) >= limit:
                        return docs[:limit]
        return docs[:limit]

File: frontend/lib/connectors/reddit.ts
import { FetchedReview, SourceRow } from "./types";

const USER_AGENT = "aisle-discovery-engine/0.2 (contact via GitHub repo)";

async function getAppOnlyToken(): Promise<string | null> {
  const id = process.env.REDDIT_CLIENT_ID;
  const secret = process.env.REDDIT_CLIENT_SECRET;
  if (!id || !secret) return null;
  const res = await fetch("https://www.reddit.com/api/v1/access_token", {
    method: "POST",
    headers: {
      Authorization: `Basic ${Buffer.from(`${id}:${secret}`).toString("base64")}`,
      "Content-Type": "application/x-www-form-urlencoded",
      "User-Agent": USER_AGENT,
    },
    body: "grant_type=client_credentials",
  });
  if (!res.ok) return null;
  const json = await res.json();
  return json.access_token ?? null;
}

/** Reddit discussions — uses OAuth app-only credentials if REDDIT_CLIENT_ID/
 * SECRET are set (free to create at reddit.com/prefs/apps, "script" type),
 * otherwise falls back to Reddit's public unauthenticated search endpoint,
 * which works for read-only access but is more likely to be rate-limited. */
export async function fetchRedditPosts(source: SourceRow, limit: number): Promise<FetchedReview[]> {
  const subreddits: string[] = source.config_json.subreddits || [];
  const terms: string[] = source.config_json.query_terms || [];
  if (subreddits.length === 0 || terms.length === 0) return [];

  const token = await getAppOnlyToken();
  const base = token ? "https://oauth.reddit.com" : "https://www.reddit.com";
  const headers: Record<string, string> = { "User-Agent": USER_AGENT };
  if (token) headers.Authorization = `Bearer ${token}`;

  const perTerm = Math.max(1, Math.ceil(limit / (subreddits.length * terms.length)));
  const reviews: FetchedReview[] = [];

  for (const sub of subreddits) {
    for (const term of terms) {
      const url = `${base}/r/${sub}/search.json?q=${encodeURIComponent(term)}&restrict_sr=1&sort=new&limit=${perTerm}`;
      try {
        const res = await fetch(url, { headers });
        if (!res.ok) continue;
        const json = await res.json();
        const children = json?.data?.children || [];
        for (const c of children) {
          const p = c.data;
          reviews.push({
            externalId: p.id,
            rawText: `${p.title || ""}\n\n${p.selftext || ""}`.trim(),
            author: p.author || "deleted",
            rating: null,
            postedAt: p.created_utc ? new Date(p.created_utc * 1000) : null,
            url: p.permalink ? `https://reddit.com${p.permalink}` : null,
            meta: { subreddit: sub, score: p.score, kind: "submission" },
          });
        }
      } catch {
        continue; // one bad subreddit/term must not kill the whole sync
      }
      if (reviews.length >= limit) return reviews.slice(0, limit);
    }
  }
  return reviews.slice(0, limit);
}
